app/estimate_square_size.go

```diff
@@ -0,0 +1,246 @@
+		// follow the non-interactive defaults are simple to estimate.
+		fits, msgShares = shares.FitsInSquare(txShares+evdShares, squareSize, msgLens...)
+		switch {
+		// stop estimating if we know we can reach the max square size
+		case squareSize >= consts.MaxSquareSize:
+			return consts.MaxSquareSize, txShares + evdShares + msgShares
+		// return if we've found a square size that fits all of the txs
+		case fits:
+			return uint64(squareSize), txShares + evdShares + msgShares
+		// try the next largest square size if we can't fit all the txs
+		case !fits:
+			// increment the square size
+			squareSize = nextPowerOfTwo(squareSize + 1)
+		}
+	}
+}
+
+// rawShareCount calculates the number of shares taken by all of the included
+// txs, evidence, and each msg. msgLens is a slice of the number of shares used
+// by each message without accounting for the non-interactive message layout
+// rules.
+func rawShareCount(txs []*parsedTx, evd core.EvidenceList) (txShares, evdShares int, msgLens []int) {
+	// msgSummary is used to keep track of the size and the namespace so that we
+	// can sort the namespaces before returning.
+	type msgSummary struct {
+		size      int
+		namespace []byte
+	}
+
+	var msgSummaries []msgSummary
+
+	// we use bytes instead of shares for tx and evd as they are encoded
+	// contiguously in the square, unlike msgs where each of which is assigned their
+	// own set of shares
+	txBytes, evdBytes := 0, 0
+	for _, pTx := range txs {
+		// if there is no wire message in this tx, then we can simply add the
+		// bytes and move on.
+		if pTx.msg == nil {
+			txBytes += len(pTx.rawTx)
+			continue
+		}
+
+		// if the there is a malleated tx, then we want to also account for the
+		// txs that gets included onchain. The formula used here over
+		// compensates for the actual size of the message, and in some cases can
+		// result in some wasted square space or picking a square size that is
+		// too large. TODO: improve by making a more accurate estimation formula
+		txBytes += overEstimateMalleatedTxSize(len(pTx.rawTx), len(pTx.msg.Message), len(pTx.msg.MessageShareCommitment))
+
+		msgSummaries = append(msgSummaries, msgSummary{shares.MsgSharesUsed(int(pTx.msg.MessageSize)), pTx.msg.MessageNameSpaceId})
+	}
+
+	txShares = txBytes / consts.TxShareSize
+	if txBytes > 0 {
+		txShares++ // add one to round up
+	}
+	// todo: stop rounding up. Here we're rounding up because the calculation for
+	// tx bytes isn't perfect. This catches those edge cases where we
+	// estimate the exact number of shares in the square, when in reality we're
+	// one byte over the number of shares in the square size. This will also cause
+	// blocks that are one square size too big instead of being perfectly snug.
+	// The estimation must be perfect or greater than what the square actually
+	// ends up being.
+	if txShares > 0 {
+		txShares++
+	}
+
+	for _, e := range evd.Evidence {
+		evdBytes += e.Size() + shares.DelimLen(uint64(e.Size()))
```

Comment:
[no change needed b/c I don't have a concrete proposal] I'm concerned this code may not be updated when compact share encoding code is modified.

Reply:
referenced in #700